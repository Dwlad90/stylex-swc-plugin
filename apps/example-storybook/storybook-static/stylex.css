
@layer priority1, priority2, priority3, priority4;
@layer priority1{
.padding-x1bg5miv{padding:12px 24px}
.padding-x1tamke2{padding:16px}
.padding-x1arfzav{padding:6px 12px}
.padding-x1ff1495{padding:8px 16px}
}
@layer priority2{
.borderColor-x7kbyhx{border-color:#ddd}
.borderColor-x159w2ql{border-color:#e0e0e0}
.borderColor-x9r1u3d{border-color:transparent}
.borderRadius-x1kogg8i{border-radius:6px}
.borderRadius-xur7f20{border-radius:8px}
.borderStyle-xng3xce{border-style:none}
.borderStyle-x1y0btm7{border-style:solid}
.borderWidth-xc342km{border-width:0}
.borderWidth-xmkeg23{border-width:1px}
.textDecoration-x1hl2dhg{text-decoration:none}
.transition-xurpv0g{transition:all.2s ease}
}
@layer priority3{
.alignItems-x6s0dn4{align-items:center}
.backgroundColor-x11jye9f{background-color:#06c}
.backgroundColor-x1596o1h{background-color:#dc3545}
.backgroundColor-x6cgz0f{background-color:#f5f5f5}
.backgroundColor-x12peec7{background-color:white}
.boxShadow-x1mh2kpm{box-shadow:0 2px 4px rgba(0,0,0,.1)}
.boxShadow-xz3rqps{box-shadow:0 4px 8px rgba(0,0,0,.15)}
.boxShadow-x1gnnqk1{box-shadow:none}
.color-x1525slw{color:#333}
.color-x33gn0b{color:#666}
.color-x1awj2ng{color:white}
.cursor-x1ypdohk{cursor:pointer}
.display-x3nfvp2{display:inline-flex}
.fontFamily-xmcqa5u{font-family:-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,sans-serif}
.fontSize-xfifm61{font-size:12px}
.fontSize-xif65rj{font-size:14px}
.fontSize-x1j61zf2{font-size:16px}
.fontSize-xosj86m{font-size:18px}
.fontWeight-xk50ysn{font-weight:500}
.fontWeight-x117nqv4{font-weight:bold}
.justifyContent-xl56j7k{justify-content:center}
.lineHeight-x1evy7pa{line-height:1.5}
.transform-xnn1q72{transform:translateY(0)}
.backgroundColor-xkdfp31:hover{background-color:#0052a3}
.backgroundColor-x37qbrc:hover{background-color:#c82333}
.backgroundColor-x1jmbrsh:hover{background-color:#e8e8e8}
.boxShadow-x1yrdse8:hover{box-shadow:0 4px 8px rgba(0,0,0,.12)}
.transform-x14ow2ja:hover{transform:translateY(-1px)}
.transform-x3kbueh:active{transform:translateY(0)}
}
@layer priority4{
.marginBottom-x1e56ztr{margin-bottom:8px}
.maxWidth-xw5ewwj{max-width:300px}
.minHeight-xe0p6wg{min-height:28px}
.minHeight-xu0aao5{min-height:36px}
.minHeight-x1gg8mnh{min-height:44px}
}