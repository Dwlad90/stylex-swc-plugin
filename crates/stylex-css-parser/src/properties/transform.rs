/*!
CSS Transform property parsing.

Handles transform property syntax with multiple transform functions.
*/

use crate::{css_types::transform_function::TransformFunction, token_parser::TokenParser};
use std::fmt::{self, Display};

/// CSS transform property value
#[derive(Debug, Clone, PartialEq)]
pub struct Transform {
  pub value: Vec<TransformFunction>,
}

impl Transform {
  /// Create a new Transform
  pub fn new(value: Vec<TransformFunction>) -> Self {
    Self { value }
  }

  pub fn parser() -> TokenParser<Transform> {
    use crate::token_parser::tokens;

    TokenParser::one_or_more_separated_by(TransformFunction::parse(), tokens::whitespace())
      .map(Transform::new, Some("to_transform"))
  }
}

impl Display for Transform {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    let function_strings: Vec<String> = self.value.iter().map(|func| func.to_string()).collect();
    write!(f, "{}", function_strings.join(" "))
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::css_types::{transform_function::*, Angle, Length, LengthPercentage, Percentage};

  #[test]
  fn test_transform_function_creation() {
    let translate = TransformFunction::TranslateAxis(TranslateAxis::new(
      LengthPercentage::Length(Length::new(10.0, "px".to_string())),
      Axis::X,
    ));

    // Test that it can be created without issues
    assert!(matches!(translate, TransformFunction::TranslateAxis(_)));
  }

  #[test]
  fn test_transform_function_display() {
    let translate_x = TransformFunction::TranslateAxis(TranslateAxis::new(
      LengthPercentage::Length(Length::new(10.0, "px".to_string())),
      Axis::X,
    ));
    assert_eq!(translate_x.to_string(), "translateX(10px)");

    let translate = TransformFunction::Translate(Translate::new(
      LengthPercentage::Length(Length::new(10.0, "px".to_string())),
      Some(LengthPercentage::Length(Length::new(
        20.0,
        "px".to_string(),
      ))),
    ));
    assert_eq!(translate.to_string(), "translate(10px, 20px)");

    let rotate = TransformFunction::Rotate(Rotate::new(Angle::new(45.0, "deg".to_string())));
    assert_eq!(rotate.to_string(), "rotate(45deg)");
  }

  #[test]
  fn test_transform_creation() {
    let func1 = TransformFunction::TranslateAxis(TranslateAxis::new(
      LengthPercentage::Length(Length::new(10.0, "px".to_string())),
      Axis::X,
    ));
    let func2 = TransformFunction::Rotate(Rotate::new(Angle::new(45.0, "deg".to_string())));

    let transform = Transform::new(vec![func1, func2]);
    assert_eq!(transform.value.len(), 2);
  }

  #[test]
  fn test_transform_display() {
    let func1 = TransformFunction::TranslateAxis(TranslateAxis::new(
      LengthPercentage::Length(Length::new(10.0, "px".to_string())),
      Axis::X,
    ));
    let func2 = TransformFunction::Rotate(Rotate::new(Angle::new(45.0, "deg".to_string())));

    let transform = Transform::new(vec![func1, func2]);
    assert_eq!(transform.to_string(), "translateX(10px) rotate(45deg)");
  }

  #[test]
  fn test_transform_empty() {
    let transform = Transform::new(vec![]);
    assert_eq!(transform.to_string(), "");
  }

  #[test]
  fn test_transform_equality() {
    let func = TransformFunction::ScaleAxis(ScaleAxis::new(1.5, Axis::X));

    let transform1 = Transform::new(vec![func.clone()]);
    let transform2 = Transform::new(vec![func]);

    assert_eq!(transform1, transform2);
  }

  #[test]
  fn test_transform_parser_creation() {
    // Test that the parser can be created without panicking
    let _parser = Transform::parser();
  }

  #[test]
  fn test_transform_single_function() {
    let transform = Transform::new(vec![
      TransformFunction::Translate(Translate::new(
        LengthPercentage::Percentage(Percentage::new(50.0)),
        Some(LengthPercentage::Percentage(Percentage::new(-50.0))),
      )),
      TransformFunction::Rotate(Rotate::new(Angle::new(45.0, "deg".to_string()))),
      TransformFunction::Scale(Scale::new(1.2, Some(1.2))),
    ]);

    assert_eq!(transform.value.len(), 3);
    assert!(transform.to_string().contains("translate(50%, -50%)"));
    assert!(transform.to_string().contains("rotate(45deg)"));
    assert!(transform.to_string().contains("scale(1.2, 1.2)"));
  }

  #[test]
  fn test_transform_complex_combination() {
    let func1 = TransformFunction::Translate(Translate::new(
      LengthPercentage::Length(Length::new(10.0, "px".to_string())),
      Some(LengthPercentage::Length(Length::new(
        20.0,
        "px".to_string(),
      ))),
    ));
    let func2 = TransformFunction::Rotate(Rotate::new(Angle::new(45.0, "deg".to_string())));
    let func3 = TransformFunction::Scale(Scale::new(1.2, Some(1.2)));

    let transform = Transform::new(vec![func1, func2, func3]);
    assert_eq!(transform.value.len(), 3);
    assert_eq!(
      transform.to_string(),
      "translate(10px, 20px) rotate(45deg) scale(1.2, 1.2)"
    );
  }

  #[test]
  fn test_transform_common_functions() {
    // Test various common transform functions
    let translate = TransformFunction::Translate(Translate::new(
      LengthPercentage::Length(Length::new(10.0, "px".to_string())),
      Some(LengthPercentage::Length(Length::new(
        20.0,
        "px".to_string(),
      ))),
    ));

    let rotate = TransformFunction::Rotate(Rotate::new(Angle::new(45.0, "deg".to_string())));

    let scale = TransformFunction::Scale(Scale::new(1.5, Some(1.5)));

    let skew = TransformFunction::Skew(Skew::new(
      Angle::new(10.0, "deg".to_string()),
      Some(Angle::new(20.0, "deg".to_string())),
    ));

    let functions = vec![translate, rotate, scale, skew];
    let transform = Transform::new(functions);

    assert_eq!(transform.value.len(), 4);
    assert!(transform.to_string().contains("translate"));
    assert!(transform.to_string().contains("rotate"));
    assert!(transform.to_string().contains("scale"));
    assert!(transform.to_string().contains("skew"));
  }
}
