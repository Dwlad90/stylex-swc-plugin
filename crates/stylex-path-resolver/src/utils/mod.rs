use std::path::{Path, PathBuf};

use path_clean::PathClean;

pub(crate) fn contains_subpath(path: &Path, sub_path: &Path) -> bool {
  path
    .display()
    .to_string()
    .split("/")
    .any(|part| part == sub_path.display().to_string())
}
pub fn relative_path(file_path: &Path, root: &Path) -> PathBuf {
  pathdiff::diff_paths(file_path, root)
    .unwrap_or_else(|| {
      panic!(
        "Failed to get relative path for file {} based on root {}",
        file_path.display(),
        root.display()
      )
    })
    .clean()
}
