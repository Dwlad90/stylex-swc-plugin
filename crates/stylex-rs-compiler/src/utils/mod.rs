use napi::{Env, Error, JsObject};
use stylex_shared::StyleXTransform;
use swc_core::plugin::proxies::PluginCommentsProxy;

pub(crate) fn extract_stylex_metadata(
  env: Env,
  stylex: &StyleXTransform<PluginCommentsProxy>,
) -> Result<Vec<JsObject>, Error> {
  let mut stylex_metadata: Vec<JsObject> = vec![];

  for (_, value) in &stylex.state.metadata {
    for meta in value.iter() {
      let mut metadata_value = env.create_array_with_length(3)?;

      metadata_value.set_element(0, env.create_string(meta.get_class_name())?)?;

      let mut style_value = env.create_object()?;

      let styles = meta.get_style();

      style_value.set_named_property("ltr", styles.ltr.clone())?;
      style_value.set_named_property("rtl", styles.rtl.clone())?;

      metadata_value.set_element(1, style_value)?;

      metadata_value.set_element(2, env.create_double(*meta.get_priority())?)?;

      stylex_metadata.push(metadata_value);
    }
  }

  Ok(stylex_metadata)
}
