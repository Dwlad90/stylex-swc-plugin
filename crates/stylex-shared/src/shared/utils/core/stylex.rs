use crate::{
  shared::{
    enums::data_structures::fn_result::FnResult, utils::ast::convertors::string_to_expression,
  },
  transform::styleq::common::styleq,
};

use super::parse_nullable_style::ResolvedArg;

pub(crate) fn stylex(values: &Vec<ResolvedArg>) -> Option<FnResult> {
  let result = styleq(values);

  Some(FnResult::Stylex(string_to_expression(
    result.class_name.as_str(),
  )))
}
