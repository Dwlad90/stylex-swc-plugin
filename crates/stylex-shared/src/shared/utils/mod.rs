pub mod ast;
pub mod common;
pub mod core;
pub mod css;
pub mod js;
pub mod log;
pub mod object;
pub(crate) mod pre_rule;
pub(crate) mod validators;
pub(crate) mod vector;
pub(crate) mod when;
