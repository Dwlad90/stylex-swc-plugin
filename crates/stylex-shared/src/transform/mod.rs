use rustc_hash::FxHashSet;
use swc_core::{
  common::{Mark, comments::Comments},
  ecma::{
    ast::{CallExpr, Callee, Expr, Id, MemberProp, Pass, VarDeclarator},
    transforms::base::resolver,
    utils::drop_span,
    visit::fold_pass,
  },
};

use crate::{
  StyleXOptionsParams,
  shared::{
    enums::core::TransformationCycle,
    structures::{
      named_import_source::{ImportSources, RuntimeInjection},
      plugin_pass::PluginPass,
      state_manager::StateManager,
      stylex_options::{CheckModuleResolution, StyleXOptions},
    },
    utils::common::increase_ident_count,
  },
};

mod fold;
pub(crate) mod styleq;
pub(crate) mod stylex;

pub struct StyleXTransform<C>
where
  C: Comments,
{
  pub comments: C,
  props_declaration: Option<Id>,
  pub state: StateManager,
}

impl<C> StyleXTransform<C>
where
  C: Comments,
{
  pub fn new(comments: C, plugin_pass: PluginPass, config: &mut StyleXOptionsParams) -> Self {
    let stylex_imports = fill_stylex_imports(&Some(config));

    let mut state = StateManager::new(config.clone().into());

    state.stylex_import.clone_from(&stylex_imports);

    state.options.import_sources = stylex_imports.into_iter().collect();

    state._state = plugin_pass;

    StyleXTransform {
      comments,
      props_declaration: None,
      state,
    }
  }

  pub fn new_test_force_runtime_injection(
    comments: C,
    plugin_pass: PluginPass,
    config: Option<&mut StyleXOptionsParams>,
  ) -> Self {
    let stylex_imports = fill_stylex_imports(&config);

    let mut state = match config {
      Some(config) => {
        config.runtime_injection = Some(true);
        config.treeshake_compensation = Some(true);

        StateManager::new(config.clone().into())
      }
      None => {
        let config = StyleXOptions {
          runtime_injection: RuntimeInjection::Boolean(true),
          treeshake_compensation: Some(true),
          unstable_module_resolution: CheckModuleResolution::Haste(
            StyleXOptions::get_haste_module_resolution(None),
          ),
          ..Default::default()
        };

        StateManager::new(config)
      }
    };

    state.options.import_sources = stylex_imports.into_iter().collect();

    state._state = plugin_pass;

    StyleXTransform {
      comments,
      props_declaration: None,
      state,
    }
  }

  pub fn new_test_force_runtime_injection_with_pass(
    comments: C,
    plugin_pass: PluginPass,
    config: Option<&mut StyleXOptionsParams>,
  ) -> impl Pass + use<C> {
    (
      resolve_factory(),
      fold_pass(Self::new_test_force_runtime_injection(
        comments,
        plugin_pass,
        config,
      )),
    )
  }

  pub fn new_test(
    comments: C,
    plugin_pass: PluginPass,
    config: Option<&mut StyleXOptionsParams>,
  ) -> Self {
    let stylex_imports = fill_stylex_imports(&config);

    let mut state = match config {
      Some(config) => StateManager::new(config.clone().into()),
      None => {
        let config = StyleXOptions {
          runtime_injection: RuntimeInjection::Boolean(false),
          treeshake_compensation: Some(true),
          class_name_prefix: "x".to_string(),
          unstable_module_resolution: CheckModuleResolution::Haste(
            StyleXOptions::get_haste_module_resolution(None),
          ),
          ..Default::default()
        };

        StateManager::new(config)
      }
    };

    state.options.import_sources = stylex_imports.into_iter().collect();

    state._state = plugin_pass;

    StyleXTransform {
      comments,
      props_declaration: None,
      state,
    }
  }
  pub fn new_test_with_pass(
    comments: C,
    plugin_pass: PluginPass,
    config: Option<&mut StyleXOptionsParams>,
  ) -> impl Pass + use<C> {
    (
      resolve_factory(),
      fold_pass(Self::new_test(comments, plugin_pass, config)),
    )
  }

  pub(crate) fn process_declaration(&mut self, call_expr: &mut CallExpr) -> Option<(Id, String)> {
    let state = &self.state;

    let stylex_imports = state.stylex_import_stringified();
    if let Callee::Expr(callee) = &mut call_expr.callee {
      match callee.as_ref() {
        Expr::Ident(ident) => {
          let ident_id = ident.to_id();

          if stylex_imports.contains(&ident.sym.to_string())
            || (state.cycle == TransformationCycle::TransformEnter
              && (state.stylex_create_import.contains(&ident.sym))
              || state.stylex_props_import.contains(&ident.sym)
              || state.stylex_keyframes_import.contains(&ident.sym)
              || state.stylex_first_that_works_import.contains(&ident.sym)
              || state.stylex_types_import.contains(&ident.sym)
              || state.stylex_create_theme_import.contains(&ident.sym)
              || state.stylex_define_vars_import.contains(&ident.sym)
              || state.stylex_attrs_import.contains(&ident.sym))
          {
            increase_ident_count(&mut self.state, ident);

            return Some((ident_id.clone(), format!("{}", ident.sym)));
          }
        }
        Expr::Member(member) => {
          if let Expr::Ident(ident) = member.obj.as_ref() {
            let ident_id = ident.to_id();

            if stylex_imports.contains(&ident.sym.to_string())
              || (state.cycle == TransformationCycle::TransformEnter
                && (state.stylex_create_import.contains(&ident.sym))
                || state.stylex_props_import.contains(&ident.sym)
                || state.stylex_keyframes_import.contains(&ident.sym)
                || state.stylex_first_that_works_import.contains(&ident.sym)
                || state.stylex_create_theme_import.contains(&ident.sym)
                || state.stylex_types_import.contains(&ident.sym)
                || state.stylex_define_vars_import.contains(&ident.sym)
                || state.stylex_attrs_import.contains(&ident.sym))
            {
              if let MemberProp::Ident(ident) = &member.prop {
                return Some((ident_id.clone(), format!("{}", ident.sym)));
              }
            }
          }
        }
        _ => {}
      }
    }

    None
  }

  pub(crate) fn transform_call_expression(&mut self, expr: &mut Expr) -> Option<Expr> {
    if let Expr::Call(ex) = expr {
      let declaration = self.process_declaration(ex);

      if declaration.is_some() {
        return self.transform_call_expression_to_stylex_expr(ex);
      }
    }

    None
  }

  pub(crate) fn get_call_var_name(
    &mut self,
    call: &CallExpr,
  ) -> (Option<String>, Option<VarDeclarator>) {
    let mut var_name: Option<String> = None;

    let parent_var_decl = self
      .state
      .declarations
      .iter()
      .find(|decl| match &decl.init {
        Some(init) => {
          if let Expr::Call(init_call) = init.as_ref() {
            init_call == &drop_span(call.clone())
          } else {
            false
          }
        }
        _ => false,
      })
      .cloned();

    if let Some(ref parent_var_decl) = parent_var_decl {
      if let Some(ident) = parent_var_decl.name.as_ident() {
        var_name = Some(ident.sym.to_string());
      }
    }

    (var_name, parent_var_decl)
  }
}

fn fill_stylex_imports(config: &Option<&mut StyleXOptionsParams>) -> FxHashSet<ImportSources> {
  let mut stylex_imports = FxHashSet::default();

  stylex_imports.insert(ImportSources::Regular("stylex".to_string()));
  stylex_imports.insert(ImportSources::Regular("@stylexjs/stylex".to_string()));

  if let Some(stylex_imports_extends) = match config {
    Some(config) => config.import_sources.clone(),
    None => None,
  } {
    stylex_imports.extend(stylex_imports_extends)
  }

  stylex_imports
}

#[warn(clippy::extra_unused_type_parameters)]
fn resolve_factory() -> impl Pass {
  let unresolved_mark = Mark::new();
  let top_level_mark = Mark::new();

  resolver(unresolved_mark, top_level_mark, true)
}
