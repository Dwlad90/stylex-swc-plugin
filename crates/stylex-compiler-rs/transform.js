var { transform } = require('./index')

console.log(
  JSON.stringify(
    transform(`
  import stylex from '@stylexjs/stylex';
  export const styles = stylex.create({
      default: {
          backgroundColor: 'red',
          color: 'blue',
      }
  });
  `),
    null,
    2,
  ),
)

/**
 * Output:
  {
    "code": "import stylex from '@stylexjs/stylex';\nexport const styles = {\n    default: {\n        backgroundColor: \"xrkmrrc\",\n        color: \"xju2f9n\",\n        $$css: true\n    }\n};\n",
    "metadata": {
      "stylex": {
        "styles": [
          {
            "className": "xrkmrrc",
            "style": {
              "ltr": ".xrkmrrc{background-color:red}",
              "rtl": null
            },
            "priority": 3000
          },
          {
            "className": "xju2f9n",
            "style": {
              "ltr": ".xju2f9n{color:blue}",
              "rtl": null
            },
            "priority": 3000
          }
        ]
      }
    },
    "sourcemap": "{\"version\":3,\"sources\":[\"<anon>\"],\"names\":[],\"mappings\":\"AACE;AACA;;;;;;EAKG\"}"
  }
 */
