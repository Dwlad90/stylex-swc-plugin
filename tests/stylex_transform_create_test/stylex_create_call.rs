use std::cell::RefCell;
use std::io::Write;
use std::{rc::Rc, sync::Arc};

use stylex_swc_plugin::ModuleTransformVisitor;
// use swc_ecma_codegen::{text_writer::JsWriter, Emitter};

use swc_core::ecma::codegen::text_writer::JsWriter;
use swc_core::ecma::codegen::Emitter;
use swc_core::{
    common::{
        comments::Comments,
        errors::{ColorConfig, Handler},
        FileName, Globals, SourceMap,
    },
    ecma::{
        ast::{EsVersion, Module},
        parser::{lexer::Lexer, Parser, StringInput, Syntax, TsConfig},
        transforms::testing::test,
        visit::FoldWith,
    },
    plugin::proxies::PluginCommentsProxy,
};

use crate::utils::transform::{parse_js, stringify_js};

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    transforms_style_object,
    r#"
        import stylex from 'stylex';
        const styles = stylex.create({
            default: {
                backgroundColor: 'red',
                color: 'blue',
            }
        });
    "#
);

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    transforms_style_object_with_import_asterics,
    r#"
        import * as foo from 'stylex';
        const styles = foo.create({
            default: {
                backgroundColor: 'red',
                color: 'blue',
            }
        });
    "#
);

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    transforms_style_object_with_named_imports,
    r#"
        import {create} from 'stylex';
        const styles = create({
            default: {
                backgroundColor: 'red',
                color: 'blue',
            }
        });
    "#
);

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    transforms_style_object_with_custom_property,
    r#"
        import stylex from 'stylex';
        const styles = stylex.create({
            default: {
                '--background-color': 'red',
            }
        });
    "#
);

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    transforms_style_object_with_custom_property_as_value,
    r#"
        import stylex from 'stylex';
        const styles = stylex.create({
            default: {
                '--final-color': 'var(--background-color)',
            }
        });
    "#
);

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    transforms_multiple_namespaces,
    r#"
        import stylex from 'stylex';
        const styles = stylex.create({
            default: {
                backgroundColor: 'red',
            },
            default2: {
                color: 'blue',
            },
        });
    "#
);

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    does_not_transform_attr_fn_value,
    r#"
        import stylex from 'stylex';
        const styles = stylex.create({
            default: {
                content: 'attr(some-attribute)',
            },
        });
    "#
);

#[test]
fn handles_camel_cased_transition_properties() {
    let camel_cased = parse_js(
        "import stylex from 'stylex';
        const styles = stylex.create({
            default: {
                transitionProperty: 'marginTop',
            },
        });",
    );

    let kebab_cased = parse_js(
        "import stylex from 'stylex';
        const styles = stylex.create({
            default: {
                transitionProperty: 'margin-top',
            },
        });",
    );

    assert_eq!(stringify_js(&camel_cased), stringify_js(&kebab_cased));

    assert_eq!(
        stringify_js(&camel_cased),
        r#"import _inject from "@stylexjs/stylex/lib/stylex-inject";
    var _inject2 = _inject;
    import stylex from 'stylex';
    _inject2(".x1cfch2b{transition-property:margin-top}", 3000);
    "#
    );
}

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    leaves_transition_properties_of_custom_properties_alone,
    r#"
        import stylex from 'stylex';
        const styles = stylex.create({
            default: {
                transitionProperty: '--foo',
            },
        });
    "#
);

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    transforms_nested_pseudo_class_to_css,
    r#"
        import stylex from 'stylex';
        const styles = stylex.create({
            default: {
                ':hover': {
                    backgroundColor: 'red',
                    color: 'blue',
                },
            },
        });
    "#
);

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    transforms_nested_pseudo_class_within_properties_to_css,
    r#"
        import stylex from 'stylex';
        const styles = stylex.create({
            default: {
                backgroundColor: {
                    ':hover': 'red',
                },
                color: {
                    ':hover': 'blue',
                }
            },
        });
    "#
);

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    transforms_array_values_as_fallbacks,
    r#"
        import stylex from 'stylex';
            const styles = stylex.create({
            default: {
                position: ['sticky', 'fixed']
            },
        });
    "#
);

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    transforms_array_values_as_fallbacks_within_media_query,
    r#"
        import stylex from 'stylex';
        const styles = stylex.create({
            default: {
                position: {
                    default: 'fixed',
                    '@media (min-width: 768px)': ['sticky', 'fixed'],
                }
            },
        });
    "#
);

// TODO: add more vendor-prefixed properties and values
test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    transforms_properties_requiring_vendor_prefixes,
    r#"
        import stylex from 'stylex';
        const styles = stylex.create({
            default: {
                userSelect: 'none',
            },
        });
    "#
);

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    transforms_valid_shorthands,
    r#"
        import stylex from 'stylex';
        const styles = stylex.create({
            default: {
                overflow: 'hidden',
                borderStyle: 'dashed',
                borderWidth: 1
            }
        });
    "#
);

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    uses_stylex_include_correctly_with_member_expressions,
    r#"
        import stylex from 'stylex';
        export const styles = stylex.create({
            foo: {
                ...stylex.include(importedStyles.foo)
            }
        });
    "#
);

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    using_stylex_include_keeps_the_compiled_object,
    r#"
        import stylex from 'stylex';
        const styles = stylex.create({
            foo: {
                ...stylex.include(importedStyles.foo),
                color: 'red',
            }
        });
        stylex.props(styles.foo);
    "#
);

test!(
    Syntax::Typescript(TsConfig {
        tsx: true,
        ..Default::default()
    }),
    |tr| ModuleTransformVisitor::new_test_styles(tr.comments.clone(), Option::None),
    uses_stylex_first_that_works_correctly,
    r#"
        import stylex from 'stylex';
        export const styles = stylex.create({
            foo: {
                position: stylex.firstThatWorks('sticky', 'fixed'),
            }
        });
    "#
);
