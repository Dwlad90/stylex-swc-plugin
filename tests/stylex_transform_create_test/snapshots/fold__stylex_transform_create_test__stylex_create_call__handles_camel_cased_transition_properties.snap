---
source: tests/stylex_transform_create_test/stylex_create_call.rs
expression: stringify_js(&camel_cased)
---
import _inject from "@stylexjs/stylex/lib/stylex-inject";
var _inject2 = _inject;
import stylex from 'stylex';
_inject2(".x1cfch2b{transition-property:margin-top}", 3000);
